idx/memory/tag_query.go

```diff
@@ -171,247 +252,751 @@ func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
 			case NOT_EQUAL:
 				q.notEqual = append(q.notEqual, e.kv)
 			case MATCH:
-				var re *regexp.Regexp
-				if e.value != "^.+" {
-					re, err = regexp.Compile(e.value)
-					if err != nil {
-						return q, errInvalidQuery
-					}
+				re, err := compileRe(e.value)
+				if err != nil {
+					return q, errInvalidQuery
 				}
-				q.match = append(q.match, kvRe{key: e.key, value: re})
+				q.match = append(q.match, kvRe{
+					key:        e.key,
+					value:      re,
+					matchCache: &sync.Map{},
+					missCache:  &sync.Map{},
+				})
 			case NOT_MATCH:
-				var re *regexp.Regexp
-				if e.value != "^.+" {
-					re, err = regexp.Compile(e.value)
-					if err != nil {
-						return q, errInvalidQuery
-					}
+				re, err := compileRe(e.value)
+				if err != nil {
+					return q, errInvalidQuery
+				}
+				q.notMatch = append(q.notMatch, kvRe{
+					key:        e.key,
+					value:      re,
+					matchCache: &sync.Map{},
+					missCache:  &sync.Map{},
+				})
+			case PREFIX:
+				q.prefix = append(q.prefix, kv{key: e.key, value: e.value})
+			case MATCH_TAG:
+				// we only allow one query by tag
+				if q.filterTagBy != 0 {
+					return q, errInvalidQuery
```

Comment:
is it possible that this branch does not trigger, yet the query is invalid? e.g. filterTagBy was set to EQUAL (valued 0), but we don't detect it. we may want to reserve 0 for unset, and only use 1 and upwards for actual values

Reply:
`q.filterTagBy` will only be set to `MATCH_TAG` or `PREFIX_TAG`. If it is `0` then it must not have been set.